### PR TITLE
Update the KerasCV bounding box output experience

### PR DESCRIPTION
**Pull request by LukeWood**
After some piloting and user feedback it is clear that the following are true:
- ragged outputs are less confusing than receiving a tensor of all -1s
- NMS wasn't filtering confidence, this is just a bad/inconsistent UX

I propose we update the behavior to output ragged boxes by default, as well as filter confidence scores by class ID.  This will overall improve consistency and UX.

### CODE REVIEW

**Comment by jbischof** [comment]
I might lack sufficient context without a demo here. Under which circumstances are `-1`s returned?

**Comment by LukeWood** [comment]
> I might lack sufficient context without a demo here. Under which circumstances are `-1`s returned?

In order to convert a ragged dimension to Dense (for TPU compat, etc) you pad -1s on to all values that are to be ignored.  Consider two images:

image 1 has:
boxes: [[0, 0, 0.5 0.5], [0.25, 0.25, .1, .1]]
classes: [0, 1]

image 2 has:
boxes: [[0, 0, 0.5 0.35]]
classes: [0]

To make these a TPU compatible batch, you will need to make these Dense tensors before passing them to `train_step()`.  This can be done by calling:

`keras_cv.bounding_box.to_dense(max_boxes=2)`, which would pad the following boxes on to make it a dense tensor:

boxes: [[[0, 0, 0.5 0.5], [0.25, 0.25, .1, .1]], [[0, 0, 0.5 0.35], [-1, -1, -1, -1]]]
classes: [[0, 1], [0, -1]]

Notice the final box is all -1s.   Our components simply know to ignore these values, and you can re-filter them out using `keras_cv.bounding_box.to_ragged(boxes)`.  In this PR, I'm simply filtering padding boxes.

File: keras_cv/layers/object_detection/multi_class_non_max_suppression.py [hunk 1]
```diff
@@ -48,6 +50,7 @@ def __init__(
         confidence_threshold=0.5,
         max_detections=100,
         max_detections_per_class=100,
+        output_ragged=True,
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/multi_class_non_max_suppression.py:53]: IIRC we had discussed this API in the past and decided it wasn't worthwhile -- is there a reason this can't always be true?

Conversely, can it always be false? It seems like since we provide a `to_ragged` construct, we can direct users to that if they want to prune results from a model?

File: keras_cv/layers/object_detection/multi_class_non_max_suppression_test.py [hunk 2]
```diff
@@ -35,6 +35,7 @@ def decode_predictions_output_shapes():
         bounding_box_format="xyxy",
         from_logits=True,
         max_detections=100,
+        output_ragged=False,
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/multi_class_non_max_suppression_test.py:38]: If we do decide to add this API flexibility, we should test both cases.

File: keras_cv/bounding_box/mask_invalid_detections.py [hunk 3]
```diff
@@ -61,6 +61,7 @@ def mask_invalid_detections(bounding_boxes, output_ragged=False):
 
     boxes = bounding_boxes.get("boxes")
     classes = bounding_boxes.get("classes")
+    confidence = bounding_boxes.get("confidence", None)
```
Comment (ianstenbit) [inline keras_cv/bounding_box/mask_invalid_detections.py:64]: Was it just a bug that this was missing before? Or is it needed as a consequence of the change to the multi-class NMS?

I can't quite compute how it's related to the rest of the PR

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/layers/object_detection/multi_class_non_max_suppression.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/object_detection/multi_class_non_max_suppression.py:53]: We can just make it true.

Comment (LukeWood) [reply keras_cv/layers/object_detection/multi_class_non_max_suppression.py:53]: > Conversely, can it always be false? It seems like since we provide a to_ragged construct, we can direct users to that if they want to prune results from a model?

This is more just confusing to users - we have seen this in the issues already.  "Why does my model output all -1s". Its a lot clearer if it outputs no boxes that it just didn't make any detections.

File: keras_cv/layers/object_detection/multi_class_non_max_suppression_test.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/object_detection/multi_class_non_max_suppression_test.py]: Removed dense case - I feel its better to only have ragged, and we can add dense if a use case arises.

File: keras_cv/bounding_box/mask_invalid_detections.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/bounding_box/mask_invalid_detections.py:64]: Its a bug, but  I had to fix it to get the PYCOCOCallback to operate correctly (it needs to match the other two)

File: keras_cv/bounding_box/mask_invalid_detections.py [hunk 4]
```diff
@@ -73,14 +74,20 @@ def mask_invalid_detections(bounding_boxes, output_ragged=False):
 
     classes = tf.where(mask, classes, -tf.ones_like(classes))
 
+    if confidence is not None:
```
Comment (jbischof) [inline keras_cv/bounding_box/mask_invalid_detections.py:77]: Wait, so confidence filtering was just broken?

File: keras_cv/layers/object_detection/multi_class_non_max_suppression.py [hunk 5]
```diff
@@ -107,7 +107,9 @@ def call(self, box_prediction, class_prediction):
             "num_detections": valid_det,
         }
         # this is required to comply with KerasCV bounding box format.
-        return bounding_box.mask_invalid_detections(bounding_boxes)
+        return bounding_box.mask_invalid_detections(
+            bounding_boxes, output_ragged=True
```
Comment (jbischof) [inline keras_cv/layers/object_detection/multi_class_non_max_suppression.py:111]: Could we expose `output_ragged` as a param?

Comment (ianstenbit) [reply keras_cv/layers/object_detection/multi_class_non_max_suppression.py:111]: Luke had previously done this, but I left a comment about it. (Sorry for the churn!)

Basically my take is that it's sorta a weird config option to have. In general, parameters that look like `output_xyz` should be instead handled as post-processing by the caller IMO.

Feel free to disagree!

Comment (jbischof) [reply keras_cv/layers/object_detection/multi_class_non_max_suppression.py:111]: OK, fine by me

File: keras_cv/bounding_box/mask_invalid_detections.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/bounding_box/mask_invalid_detections.py:77]: Not the filtering but converting confidence tensors you ragged.  Basically confidence score still had -1s padded on - even in ragged mode

File: keras_cv/bounding_box/to_ragged_test.py [hunk 6]
```diff
@@ -23,13 +23,27 @@ def test_converts_to_ragged(self):
                 [[[0, 0, 0, 0], [0, 0, 0, 0]], [[2, 3, 4, 5], [0, 1, 2, 3]]]
             ),
             "classes": tf.constant([[-1, -1], [-1, 1]]),
+            "confidence": tf.constant([[0.5, 0.7], [0.23, 0.12]]),
```
Comment (jbischof) [inline keras_cv/bounding_box/to_ragged_test.py:26]: How is there a confidence if no class is detected? Confused

Also, why is the one positive class label have the lowest confidence.

**Review by jbischof — APPROVED** [review]
Idea is good! I'm a bit less clear on the details but maybe it's not important to dive in here

File: keras_cv/bounding_box/to_ragged_test.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/bounding_box/to_ragged_test.py:26]: Any box with a -1 class is considered padding.  These values could be anything - they’re irrelevant.

Comment (LukeWood) [reply keras_cv/bounding_box/to_ragged_test.py:26]: (One case this could happen is if there is a high IOU between a box with a high confidence, and a box with EVEN higher confidence.  The high confidence box will be pruned, yielding a class ID of -1, but a high confidence.  Filtering out using to_ragged() removes these values which is best for UX
